pep-9999.rst

```diff
@@ -0,0 +1,341 @@
+time.)
+
+For these top level imports, there are two exceptions which will make them
+eager (not lazy): imports inside ``try``/``except``/``finally``  or ``with``
+blocks, and star imports (``from foo import *``.) Imports inside
+exception-handling blocks remain eager so that any exceptions arising from the
+import can be handled. Star imports must remain eager since performing the
+import is the only way to know which names should be added to the namespace.
+
+Imports inside class definitions or inside functions/methods are not "top
+level" and are never lazy.
+
+Dynamic imports using ``__import__`` or ``importlib.import_module`` are also
+never lazy.
+
+
+Per-module opt out
+------------------
+
+Due to the backwards compatibility issues mentioned below, it may be necessary
+to force some imports to be eager. In first-party code, this can be easily
+accomplished via any module-level reference to the name, e.g. even re-assigning
+the name to itself will trigger the import:
+
+.. code-block python
+    import foo
+    
+    # ensure 'foo' is eagerly imported
+    foo = foo
+
+Another option that scales better to making multiple imports lazy is to place
+them inside a ``try/finally``:
+
+.. code-block py
+    try:  # force these imports to be eager
+        import foo
+        import bar
+    finally:
+        pass
+
+The more difficult case can occur if an import in third-party code that can't
+easily be modified must be forced to be eager. For this purpose, we propose to
+add an API to ``importlib`` that can be called early in the process to specify
+a list of module names within which all imports will be eager:
+
+.. code-block py
+    from importlib import set_eager_imports
+
+    set_eager_imports(["one.mod", "another"])
+
+
+Backwards Compatibility
+=======================
+
+This proposal preserves full backwards compatibility when the feature is
+disabled, which is the default.
+
+When enabled, lazy imports could produce currently unexpected results and
+behaviors in existing codebases. The problems that we may see when enabling
+lazy imports in an existing codebase are related to:
+
+
+Import Side Effects
+-------------------
+
+Import side effects that would otherwise be produced by the execution of
+imported modules during the execution of import statements will be deferred at
+least until the imported objects are used.
+
+These import side effects may include:
+
+* code executing any side-effecting logic during import;
+* relying on imported submodules being set as attributes in the parent module.
+
+
+Dynamic Paths
+-------------
+
+There could be issues related to dynamic Python import paths; particularly,
+adding (and then removing after the import) paths from ``sys.path``.
+
+.. code-block py
+    sys.path.insert(0, "/path/to/foo/module")
+    import foo
+    del sys.path[0]
+    foo.Bar()
+
+In this case, with lazy imports enabled, the import of ``foo`` will not
+actually occur while the addition to ``sys.path`` is present.
```

Comment:
It's probably important to discuss, as we chatted about, whether it makes sense to separate finder from loader execution.  IOW, I believe this PEP is proposing that neither finder nor loader executes when the `import` statement is lazy.  It *could* run the finder and cache those results, but defer loader execution until eager time.  I'm not saying it should, but the PEP should be explicit about this and perhaps discuss it.

Reply:
Yep, this is currently discussed below under rejected ideas.